engine-standalone-storage/src/engine_state.rs

```diff
@@ -0,0 +1,136 @@
+use aurora_engine_sdk::io::{StorageIntermediate, IO};
+use rocksdb::DB;
+use std::cell::{Cell, RefCell};
+
+use crate::diff::{Diff, DiffValue};
+use crate::StoragePrefix;
+
+#[derive(Debug)]
+pub enum EngineStorageValue<'a> {
+    Slice(&'a [u8]),
+    Vec(Vec<u8>),
+}
+
+impl<'a> AsRef<[u8]> for EngineStorageValue<'a> {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::Slice(slice) => slice,
+            Self::Vec(bytes) => bytes,
+        }
+    }
+}
+
+impl<'a> StorageIntermediate for EngineStorageValue<'a> {
+    fn len(&self) -> usize {
+        self.as_ref().len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.as_ref().is_empty()
+    }
+
+    fn copy_to_slice(&self, buffer: &mut [u8]) {
+        buffer.copy_from_slice(self.as_ref())
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct Storage<'db, 'input, 'output> {
+    input: &'input [u8],
+    bound_block_height: u64,
+    bound_tx_position: u16,
+    transaction_diff: &'output RefCell<Diff>,
+    output: &'output Cell<Vec<u8>>,
+    db: &'db DB,
+}
+
+impl<'db, 'input, 'output> Storage<'db, 'input, 'output> {
+    pub fn new(
+        input: &'input [u8],
+        bound_block_height: u64,
+        bound_tx_position: u16,
+        transaction_diff: &'output RefCell<Diff>,
+        output: &'output Cell<Vec<u8>>,
+        db: &'db DB,
+    ) -> Self {
+        Self {
+            input,
+            bound_block_height,
+            bound_tx_position,
+            transaction_diff,
+            output,
+            db,
+        }
+    }
+
+    pub fn get_transaction_diff(&self) -> Diff {
+        self.transaction_diff.borrow().clone()
+    }
```

Comment:
If we are returning a clone here, shouldn't it be stored owned?

Reply:
The `IO` instance only holds a reference to the `Diff` so that the instance can implement `Copy` (this allows us to move it between the engine and connector submodules without much overhead; some refactoring could possibly make it so this is not required, but for now it is).

We clone here so that the returned `Diff` can outlive the `IO` instance (the `Diff` will still be used after the `IO` instance is gone because we need to record the `Diff` to storage after we execute the transaction).

I don't think it makes sense to take the `Diff` from the `Storage` object directly (where it is not a reference) because semantically it is the `IO` instance that generates the `Diff`. Also this means under the hood the `Storage` implementation can reuse the same `Diff` allocation between multiple transactions, so I don't think the clone here is even that much overhead (if we somehow did not clone then we would need to still allocate new memory for the next `Diff` when we do a new transaction; here we allocate new memory to capture the `Diff` and reuse the previous memory in the next transaction).

Reply:
As we had discussed earlier, thanks for the detailed response for others that may be curious.